syntax = "proto3";

import "google/protobuf/empty.proto";

message Tensor {
    enum DataType {
        BOOL = 0;
        INT16 = 1;
        INT32 = 2;
        INT64 = 3;
        FP16 = 4;
        FP32 = 5;
        FP64 = 6;
    }
    string name = 1;
    DataType data_type = 2;
    repeated int64 dim = 3;
    bytes content = 4;
    repeated int64 indices = 5;
    int64 version = 6;
    string initializer = 7;
}

service PServer {
    rpc pull_param(Tensor) returns (Tensor) {}
    rpc push_param(Tensor) returns (google.protobuf.Empty) {}
    rpc pull_embedding_param(Tensor) returns (Tensor) {}
    rpc push_embedding_param(Tensor) returns (google.protobuf.Empty) {}
    rpc push_grad(Tensor) returns (google.protobuf.Empty) {}
    rpc push_embedding_grad(Tensor) returns (google.protobuf.Empty) {}
}

message EndPoint {
    repeated string endpoint = 1;
}

message Worker {
    int64 id = 1;
}

service Master {
    rpc get_pserver(Worker) returns (EndPoint) {}
}